### DISCUSSION TITLE
Need Help: AI Code for Financial Time Series Forecasting

### DISCUSSION THREAD

**Discussion by mounircherifikr** [community/community#158907] [Programming Help] [answered]
### Body

Hello,

I am a Master's student currently in the graduation phase, and I am working on a research project related to forecasting financial time series. I have historical financial data (such as prices or returns), and I would like to build an AI model to predict future values.
I am looking for a ready-to-use code or an applied example that could help me generate forecasts based on my historical dataset. If possible, I would greatly appreciate receiving code in Python.

Thank you very much for your support and assistance.

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by SUKUNA-AI** [comment, marked as answer]
Hello!

I suggest you to familiarize yourself with these repositories:
https://github.com/ShubhamG2311/Financial-Time-Series-Forecasting
https://github.com/BenjiKCF/Neural-Net-with-Financial-Time-Series-Data

Kaggle:
https://www.kaggle.com/code/dave4ml/financial-time-series
https://www.kaggle.com/code/konradb/ts-3-time-series-for-finance
https://www.kaggle.com/code/avikumart/timeseries-stock-price-analysis-forecasting

Recommendations for research
Since you are working on a research project, my advice is:

Start simple: Use the provided LSTM code or notebooks with Kaggle for a quick start.
Experiment with tsai: Try modern models like PatchTST for potentially more accurate results.
Explore innovative approaches: The ShubhamG2311 repository offers unique methods, such as time series to image conversion, that may be interesting to explore.
Compare models: Test multiple architectures (LSTM, CNNs, Transformers) on your data to determine the most efficient one.
Limitations and considerations
Data quality: Financial data is often noisy, so preprocessing (outlier removal, normalization) is critical.
Overtraining: Deep models can be overtrained on small datasets. Use regularization and validation.
Volatility: Financial markets are subject to sudden changes, which can reduce forecast accuracy.

I provide a sample Python code that uses LSTM to forecast financial time series. This code can be adapted to your data by replacing the file path and customizing the parameters:

import numpy as np
import pandas as pd
import torch
import torch.nn as nn
import torch.optim as optim
from sklearn.preprocessing import MinMaxScaler
import matplotlib.pyplot as plt
from torch.utils.data import Dataset, DataLoader

# Проверка доступности GPU
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print(f"Using device: {device}")

# Загрузка данных
def load_data(file_path):
    try:
        data = pd.read_csv(file_path)
        if 'Close' not in data.columns:
            raise ValueError("Column 'Close' not found in the dataset")
        # Обработка пропусков
        data['Close'] = data['Close'].fillna(method='ffill')
        prices = data['Close'].values.reshape(-1, 1)
        return prices
    except Exception as e:
        print(f"Error loading data: {e}")
        return None

# Подготовка данных для LSTM
def create_dataset(data, time_steps):
    X, y = [], []
    for i in range(len(data) - time_steps):
        X.append(data[i:i + time_steps])
        y.append(data[i + time_steps])
    return np.array(X), np.array(y)

# Создание пользовательского датасета для PyTorch
class TimeSeriesDataset(Dataset):
    def __init__(self, X, y):
        self.X = torch.tensor(X, dtype=torch.float32)
        self.y = torch.tensor(y, dtype=torch.float32)
    
    def __len__(self):
        return len(self.X)
    
    def __getitem__(self, idx):
        return self.X[idx], self.y[idx]

# Определение модели LSTM
class LSTMModel(nn.Module):
    def __init__(self, input_size=1, hidden_size=50, num_layers=2, dropout=0.2):
        super(LSTMModel, self).__init__()
        self.hidden_size = hidden_size
        self.num_layers = num_layers
        self.lstm = nn.LSTM(input_size, hidden_size, num_layers, batch_first=True, dropout=dropout)
        self.fc = nn.Linear(hidden_size, 1)
    
    def forward(self, x):
        # Инициализация скрытого состояния
        h0 = torch.zeros(self.num_layers, x.size(0), self.hidden_size).to(device)
        c0 = torch.zeros(self.num_layers, x.size(0), self.hidden_size).to(device)
        
        # Прямой проход через LSTM
        out, _ = self.lstm(x, (h0, c0))
        # Берем только последний временной шаг
        out = self.fc(out[:, -1, :])
        return out

# Основная функция для обучения и прогнозирования
def train_and_predict(file_path, time_steps=60, hidden_size=50, num_layers=2, dropout=0.2, epochs=20, batch_size=32):
    # Загрузка и предобработка данных
    prices = load_data(file_path)
    if prices is None:
        return
    
    # Нормализация данных
    scaler = MinMaxScaler()
    scaled_prices = scaler.fit_transform(prices)
    
    # Подготовка данных
    X, y = create_dataset(scaled_prices, time_steps)
    
    # Разделение на обучающую и тестовую выборки
    train_size = int(len(X) * 0.8)
    X_train, X_test = X[:train_size], X[train_size:]
    y_train, y_test = y[:train_size], y[train_size:]
    
    # Создание датасетов и загрузчиков
    train_dataset = TimeSeriesDataset(X_train, y_train)
    test_dataset = TimeSeriesDataset(X_test, y_test)
    train_loader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)
    test_loader = DataLoader(test_dataset, batch_size=batch_size, shuffle=False)
    
    # Инициализация модели
    model = LSTMModel(hidden_size=hidden_size, num_layers=num_layers, dropout=dropout).to(device)
    criterion = nn.MSELoss()
    optimizer = optim.Adam(model.parameters(), lr=0.001)
    
    # Обучение модели
    model.train()
    for epoch in range(epochs):
        total_loss = 0
        for X_batch, y_batch in train_loader:
            X_batch, y_batch = X_batch.to(device), y_batch.to(device)
            optimizer.zero_grad()
            output = model(X_batch)
            loss = criterion(output, y_batch)
            loss.backward()
            optimizer.step()
            total_loss += loss.item()
        print(f"Epoch {epoch+1}/{epochs}, Loss: {total_loss/len(train_loader):.6f}")
    
    # Оценка модели
    model.eval()
    predictions = []
    actuals = []
    with torch.no_grad():
        for X_batch, y_batch in test_loader:
            X_batch = X_batch.to(device)
            output = model(X_batch)
            predictions.extend(output.cpu().numpy())
            actuals.extend(y_batch.cpu().numpy())
    
    # Обратное преобразование данных
    predictions = scaler.inverse_transform(np.array(predictions))
    actuals = scaler.inverse_transform(np.array(actuals))
    
    # Вычисление RMSE
    rmse = np.sqrt(np.mean((predictions - actuals) ** 2))
    print(f"Test RMSE: {rmse:.4f}")
    
    # Визуализация результатов
    plt.figure(figsize=(12, 6))
    plt.plot(actuals, label='Actual Prices')
    plt.plot(predictions, label='Predicted Prices')
    plt.title('LSTM Predictions vs Actual Prices')
    plt.xlabel('Time')
    plt.ylabel('Price')
    plt.legend()
    plt.savefig('lstm_predictions.png')
    plt.close()
    
    # Сохранение модели
    torch.save(model.state_dict(), 'lstm_model.pth')
    
    return model, scaler

# Пример использования
if __name__ == "__main__":
    file_path = 'your_data.csv'  # Замените на путь к вашему файлу
    model, scaler = train_and_predict(file_path, time_steps=60, hidden_size=50, num_layers=2, dropout=0.2, epochs=20, batch_size=32)

**Reply by mounircherifikr** [reply]
Thanks a ton for the code. Really appreciate it.

**Comment by agiletechvietnam1** [comment]
Certainly! For your research on forecasting financial time series using AI, here are some valuable resources and suggestions:

---

### 📘 Recommended Resources

**GitHub Repositories:**

* [[Financial-Time-Series-Forecasting](https://github.com/ShubhamG2311/Financial-Time-Series-Forecasting)](https://github.com/ShubhamG2311/Financial-Time-Series-Forecasting): Offers practical examples of time series forecasting using various models.

* [[Neural-Net-with-Financial-Time-Series-Data](https://github.com/BenjiKCF/Neural-Net-with-Financial-Time-Series-Data)](https://github.com/BenjiKCF/Neural-Net-with-Financial-Time-Series-Data): Demonstrates the application of neural networks to financial data.

**Kaggle Notebooks:**

* [[Financial Time Series Forecasting](https://www.kaggle.com/code/dave4ml/financial-time-series)](https://www.kaggle.com/code/dave4ml/financial-time-series): Provides a comprehensive approach to financial time series forecasting.

* [[TS-3: Time Series for Finance](https://www.kaggle.com/code/konradb/ts-3-time-series-for-finance)](https://www.kaggle.com/code/konradb/ts-3-time-series-for-finance): Focuses on time series analysis in the financial sector.

* [[Stock Price Analysis & Forecasting](https://www.kaggle.com/code/avikumart/timeseries-stock-price-analysis-forecasting)](https://www.kaggle.com/code/avikumart/timeseries-stock-price-analysis-forecasting): Explores stock price analysis and forecasting techniques.

---

### 🛠️ Getting Started

1. **Data Preparation:**

   * Ensure your historical financial data is cleaned and preprocessed.
   * Handle missing values and outliers appropriately.

2. **Model Selection:**

   * Start with traditional models like ARIMA or Exponential Smoothing for baseline comparisons.
   * Explore machine learning models such as Random Forests or Gradient Boosting.
   * Delve into deep learning models like LSTM or GRU for capturing complex patterns.

3. **Evaluation:**

   * Use metrics like RMSE, MAE, or MAPE to assess model performance.
   * Perform cross-validation to ensure model robustness.

You can find more information in here: https://agiletech.vn/technologies/

**Comment by NaeemAbdullahAkram** [comment]
> `pip install u8darts[torch] pandas matplotlib
`
`import pandas as pd
import matplotlib.pyplot as plt
from darts import TimeSeries
from darts.models import TransformerModel
from darts.metrics import mape

# Load dataset (example: historical stock prices)
df = pd.read_csv("your_data.csv", parse_dates=["Date"])

# Ensure columns: 'Date' and 'Close' or adjust accordingly
series = TimeSeries.from_dataframe(df, time_col="Date", value_cols="Close")

# Normalize (optional but recommended)
series = series.normalize()

# Train/test split (e.g., last 20% as test)
train, val = series.split_before(0.8)

# Initialize Transformer model
model = TransformerModel(
    input_chunk_length=30,       # how many past days to use
    output_chunk_length=7,       # how many days to forecast
    batch_size=16,
    n_epochs=100,
    model_name="transformer_timeseries",
    nr_epochs_val_period=1,
    random_state=42,
    likelihood=None,
    verbose=True,
)

# Fit the model
model.fit(train, val_series=val, verbose=True)

# Forecast next 7 steps
forecast = model.predict(n=7)

# Plotting
series.plot(label="actual")
forecast.plot(label="forecast")
plt.legend()
plt.title("Financial Time Series Forecast (Transformer)")
plt.show()

# Evaluation
print("MAPE:", mape(val[-7:], forecast))
`

**Comment by NaeemAbdullahAkram** [comment]
That's a great project, and there are several high-quality resources with ready-to-use Python code for financial time series forecasting using AI. Here are the best ones categorized by approach:

🧠 Machine Learning-Based Forecasting
[Time Series Forecasting with Scikit-learn (GitHub)](https://github.com/WillKoehrsen/Data-Analysis/tree/master/time%20series)

ML techniques: Random Forests, Gradient Boosting

Clean, well-commented code

Good for benchmarking classical ML

Time Series Forecasting with XGBoost and LightGBM

Step-by-step tutorial

Shows how to convert time series into a supervised learning problem

🧠📈 Deep Learning-Based Forecasting
Time Series Forecasting with LSTM using Keras

Uses LSTM for financial forecasting

Based on Keras/TensorFlow

Excellent for beginners to intermediate DL users

[PyTorch Time Series Example (GitHub - Sales forecasting)](https://github.com/PacktPublishing/Deep-Learning-for-Time-Series-Forecasting)

PyTorch-based examples

Also includes CNNs, RNNs, and Transformers

Part of a great book: Deep Learning for Time Series Forecasting

🧰 Ready Libraries & Frameworks
[Darts (by Unit8)](https://github.com/unit8co/darts)

State-of-the-art forecasting library

Includes ARIMA, Prophet, RNN, LSTM, Transformer models

Easy to plug in your financial data

[Facebook Prophet](https://github.com/facebook/prophet)

Very intuitive for time series forecasting

Good for daily or seasonal financial data

Plug-and-play Python interface
